hpx/runtime/serialization/array.hpp

```diff
@@ -0,0 +1,205 @@
+                serialize_optimized(ar, v, boost::mpl::false_());
+            else
+                serialize_optimized(ar, v, use_optimized());
+        }
+
+    private:
+        value_type* m_t;
+        std::size_t m_element_count;
+    };
+
+    template <>
+    template <class Archive>
+    void array<float>::serialize(Archive& ar, unsigned int v)
+    {
+        serialize_optimized(ar, v, boost::mpl::true_());
+    }
+
+    template <>
+    template <class Archive>
+    void array<double>::serialize(Archive& ar, unsigned int v)
+    {
+        serialize_optimized(ar, v, boost::mpl::true_());
+    }
+
+    template <>
+    template <class Archive>
+    void array<char>::serialize(Archive& ar, unsigned int v)
+    {
+        serialize_optimized(ar, v, boost::mpl::true_());
+    }
+
+    template <>
+    template <class Archive>
+    void array<signed char>::serialize(Archive& ar, unsigned int v)
+    {
+        serialize_optimized(ar, v, boost::mpl::true_());
+    }
+
+    template <>
+    template <class Archive>
+    void array<unsigned char>::serialize(Archive& ar, unsigned int v)
+    {
+        serialize_optimized(ar, v, boost::mpl::true_());
+    }
+
+    template <>
+    template <class Archive>
+    void array<bool>::serialize(Archive& ar, unsigned int v)
+    {
+        serialize_optimized(ar, v, boost::mpl::true_());
+    }
+
+    // make_array function
+    template <class T> BOOST_FORCEINLINE
+    array<T> make_array(T* begin, std::size_t size)
+    {
+        return array<T>(begin, size);
+    }
+
+    // implement serialization for boost::array
+    template <class Archive, class T, std::size_t N>
+    void serialize(Archive& ar, boost::array<T,N>& a, const unsigned int /* version */)
+    {
+        ar & hpx::serialization::make_array(a.begin(), a.size());
+    }
+
+#ifndef BOOST_NO_CXX11_HDR_ARRAY
+  // implement serialization for std::array
+    template <class Archive, class T, std::size_t N>
+    void serialize(Archive& ar, std::array<T,N>& a, const unsigned int /* version */)
+    {
+        ar & hpx::serialization::make_array(a.begin(), a.size());
+    }
+#endif
+
+    // allow our array to be serialized as prvalue
+    // compiler should support good ADL implementation
+    // but it is rather for all hpx serialization library
+    template <typename T> BOOST_FORCEINLINE
+    output_archive & operator<<(output_archive & ar, array<T> t)
+    {
+        ar.invoke(t);
+        return ar;
+    }
+
+    template <typename T> BOOST_FORCEINLINE
+    input_archive & operator>>(input_archive & ar, array<T> t)
+    {
+        ar.invoke(t);
+        return ar;
+    }
+
+    template <typename T> BOOST_FORCEINLINE
+    output_archive & operator&(output_archive & ar, array<T> t)
+    {
+        ar.invoke(t);
+        return ar;
+    }
+
+    template <typename T> BOOST_FORCEINLINE
+    input_archive & operator&(input_archive & ar, array<T> t)
+    {
+        ar.invoke(t);
+        return ar;
+    }
```

Comment:
Since `array` is shallow, there shouldn't be a real need to take it by ref. If not doing so conflicts with overload resolution, it should always be possible to take them by const ref.